dali/operators/reader/loader/nemo_asr_loader.cc

```diff
@@ -0,0 +1,192 @@
+    detail::ParseManifest(entries_, fstream, max_duration_);
+  }
+
+  DALI_ENFORCE(Size() > 0, "No files found.");
+  if (shuffle_) {
+    // seeded with hardcoded value to get
+    // the same sequence on every shard
+    std::mt19937 g(kDaliDataloaderSeed);
+    std::shuffle(entries_.begin(), entries_.end(), g);
+  }
+  Reset(true);
+}
+
+void NemoAsrLoader::Reset(bool wrap_to_shard) {
+  current_index_ = wrap_to_shard ? start_index(shard_id_, num_shards_, Size()) : 0;
+  current_epoch_++;
+
+  if (shuffle_after_epoch_) {
+    std::mt19937 g(kDaliDataloaderSeed + current_epoch_);
+    std::shuffle(entries_.begin(), entries_.end(), g);
+  }
+}
+
+void NemoAsrLoader::PrepareEmpty(AsrSample &sample) {
+  PrepareEmptyTensor(sample.audio);
+}
+
+void NemoAsrLoader::ReadSample(AsrSample& sample) {
+  auto &entry = entries_[current_index_++];
+
+  // handle wrap-around
+  MoveToNextShard(current_index_);
+
+  // metadata info
+  DALIMeta meta;
+  meta.SetSourceInfo(entry.audio_filepath);
+  meta.SetSkipSample(false);
+  sample.audio.SetMeta(meta);
+
+  // Ignoring copy_read_data_, Sharing data is not supported with this loader
+
+  using DecoderType = int16_t;
+  GenericAudioDecoder<DecoderType> decoder;
+  sample.audio_meta = decoder.OpenFromFile(entry.audio_filepath);
+  assert(sample.audio_meta.channels_interleaved);  // it's always true
+
+  sample.audio.set_type(TypeTable::GetTypeInfo(dtype_));
+  auto shape = DecodedAudioShape(sample.audio_meta, sample_rate_, downmix_);
+  assert(shape.size() > 0);
+  sample.audio.Resize(shape);
+
+  bool should_resample = sample_rate_ > 0 && sample.audio_meta.sample_rate != sample_rate_;
+  bool should_downmix = sample.audio_meta.channels > 1 && downmix_;
+
+  int64_t decode_scratch_sz = 0;
+  int64_t resample_scratch_sz = 0;
+  if (should_resample || should_downmix || dtype_ != DALI_INT16)
+    decode_scratch_sz = sample.audio_meta.length * sample.audio_meta.channels;
+
+  // resample scratch is used to prepare a single or multiple (depending if
+  // downmixing is needed) channel float input, required by the resampling
+  // kernel
+  int64_t out_channels = should_downmix ? 1 : sample.audio_meta.channels;
+  if (should_resample)
+    resample_scratch_sz = sample.audio_meta.length * out_channels;
+
+  int64_t total_scratch_sz =
+      decode_scratch_sz * sizeof(DecoderType) + resample_scratch_sz * sizeof(float);
+  scratch_.set_type(TypeTable::GetTypeInfo(DALI_UINT8));
+  scratch_.Resize({total_scratch_sz});
+  uint8_t* scratch_mem = scratch_.mutable_data<uint8_t>();
+
+  span<DecoderType> decoder_scratch_mem(reinterpret_cast<DecoderType *>(scratch_mem),
+                                        decode_scratch_sz);
+  span<float> resample_scratch_mem(
+        reinterpret_cast<float *>(scratch_mem + decode_scratch_sz * sizeof(DecoderType)),
+        resample_scratch_sz);
+  TYPE_SWITCH(dtype_, type2id, OutType, (float, int16_t), (
```

Comment:
I wonder how perf looks like as Readers is single threaded and decoding + resampling could be a pretty CPU intensive task?
I wonder if we could separate away the decoding + resampling itself from opening the file itself.

Reply:
I'd prefer to keep decoding as part of the reader, as we might get a big recording that is artificially split into small chunks via the manifest file. In that case, returning raw encoded audio would be wasteful.
I'll think how to make the reader multi-threaded